public/docs/dart/latest/guide/dependency-injection.jade

```diff
@@ -1,6 +1,782 @@
+
+.l-sub-section
+  :marked
+    Angular itself uses `OpaqueTokens` to register all of its own non-class dependencies. For example,
+    [HTTP_PROVIDERS](https://angular.io/docs/ts/latest/api/http/HTTP_PROVIDERS-let.html)
+    is the `OpaqueToken` associated with an array of providers for persisting data with the Angular `Http` client.
+
+    Internally, the `Provider` turns both the string and the class type into an `OpaqueToken`
+    and keys its *token/provider* map with that.
+
+.l-main-section
+:marked
+  # Next Steps
+  We learned the basics of Angular Dependency Injection in this chapter.
+
+  The Angular Dependency Injection is more capable than we've described.
+  We can learn more about its advanced features, beginning with its support for
+  nested injectors, in the
+  [Hierarchical Dependency Injection](hierarchical-dependency-injection.html) chapter.
+
+.l-main-section
+<a id="explicit-injector"></a>
+:marked
+  ### Appendix: Working with injectors directly
+  We rarely work directly with an injector.
+  Here's an `InjectorComponent` that does.
+
++makeExample('dependency-injection/dart/web/injector.component.dart', 'injector',
+'app/injector.component.dart')
+:marked
+  The `Injector` is itself an injectable service.
+
+  In this example, Angular injects the component's own `Injector` into the component's constructor.
+  The component then asks the injected injector for the services it wants.
+
+  Note that the services themselves are not injected into the component.
+  They are retrieved by calling `injector.get`.
+
+  The `get` method throws an error if it can't resolve the requested service.
+  We can call `getOptional` instead, which we do in one case
+  to retrieve a service (`ROUS`) that isn't registered with this or any ancestor injector.
+
+.l-sub-section
+  :marked
+    The technique we just described is an example of the
+    [Service Locator Pattern](https://en.wikipedia.org/wiki/Service_locator_pattern).
+
+    We **avoid** this technique unless we genuinely need it.
+    It encourages a careless grab-bag approach such as we see here.
+    It's difficult to explain, understand, and test.
+    We can't know by inspecting the constructor what this class requires or what it will do.
+    It could acquire services from any ancestor component, not just its own.
+    We're forced to spelunk the implementation to discover what it does.
+
+    Framework developers may take this approach when they
+    must acquire services generically and dynamically.
+
+.l-main-section
+<a id="forward-ref"></a>
+:marked
+  ### Appendix: Why we recommend one class per file
+
+  Having multiple classes in the same file is confusing and best avoided.
+  Developers expect one class per file. Keep them happy.
+
+  If we scorn this advice and, say,
+  combine our `HeroService` class with the `HeroesComponent` in the same file,
+  **define the component last!**
+  If we define the component before the service,
```

Comment:
This doesn't apply to Dart. In TS this would need `forwardRef(() => {})` but in Dart the order of the classes doesn't matter.